### PR TITLE
tests and examples for the complex relative entropy cone

### PR DESCRIPTION
**Pull request by araujoms**
I got the tests to pass and the examples to run with the new cone (I get an error with "contraction", "semidefinitepoly", and "convexityparameter" but they are unrelated).

1. I have to apologize for the hack in `arrayutilities.jl`. I couldn't figure out how to get the functions `smat_to_svec!` and `svec_to_smat!` to dispatch correctly with complex Hermitian JuMP variables. I hope you can tell me how to do that so I can fix it.
2. I couldn't figure out what you were trying to do in the example "entanglementassisted", but it was giving the wrong answer anyway, so I rewrote it to something I do understand.
3. Now JuMP accepts models with arbitrary types, so I modified these three examples accordingly. I couldn't figure out, however, how to get them to actually run. I assume for you this will be trivial.
4. I've added the asymptotic expansion for the central points; it's probably useless but nice to look at.
5. I'm rather puzzled about the interface of the cone: to get the relative entropy D(W||V) I have to call the cone with the variables in the order V, W. Is there any reason for doing that? Do you mind if I change it? I wasted a lot of time debugging until I found that this was the error. Also, the order was wrong in the example "entanglementassisted". Another thing: instead of passing the dimension of the matrix as an argument, as in the PSD cone, we need to pass the total number of real parameters. Isn't that a bit inconvenient? Do you mind if I change it?

### CODE REVIEW

**Comment by araujoms** [comment]
Thanks for the answers!

2. In the meanwhile I have read [arXiv:2103.04104](https://arxiv.org/abs/2103.04104) so now I understand what you were doing. Indeed, fixing the order of the arguments was enough to make it give the right answer. Surprisingly, it turned out that computing the von Neumann entropy with `EpiTrRelEntropyTriCone` was slightly faster than with `EpiPerSepSpectralCone`, but I changed it back anyway.
3. I'm sorry, I was unclear. I know how to run the examples, what I couldn't figure out is how to pass generic types as arguments to the JuMP examples through your interface; right now the JuMP examples are only run with `Float64`.
5. I don't see how v is a perspective in the vector relative entropy cone; maybe you meant the logarithm cone instead? Also, in equation (60) of coneref.pdf you say that you are interleaving the elements of the vectors v and w, is this actually true? As for the dimension, indeed, I was confusing the MOI version of the PSD cone with your version.

**Comment by araujoms** [comment]
Could you rebase this branch on the current master, so that it works with julia 1.10?

**Comment by araujoms** [comment]
:facepalm: Yes, of course, I can just rebase it myself, I don't know what I was thinking.

**Comment by chriscoey** [comment]
Thanks for this, and sorry I've been too busy to review. Let's let @odow merge #832 and #833, then update this branch and push again and see if we can fix the failing tests.

**Comment by chriscoey** [comment]
should be ok to merge in master again for the 1.10 fixes. you will have to Pkg add [JuliaFormatter](https://github.com/domluna/JuliaFormatter.jl) before dev'ing Hypatia and then run `format(".")` in the root of the Hypatia repo, in order to fix any formatting issues and get the FormatCheck to pass. then push and we'll see what's going wrong with the examples tests.

**Comment by chriscoey** [comment]
also i'm in the process of moving the repo ownership over to the jump-dev github organization, so that it can be a bit better maintained!

**Comment by araujoms** [comment]
I understand, life after academia is tough. I've merged in master and ran JuliaFormatter, let's see how it goes now.

**Comment by chriscoey** [comment]
thank you. the format checks now pass. the failing examples seem to just be related to the sparse PSD cone. I'll have a look into those failures later today.

**Review by chriscoey — APPROVED** [review]

**Comment by chriscoey** [comment]
@odow this PR fails on one test for one cone, only on julia 1.6 
`MethodError: no method matching rdiv!(::Matrix{ComplexF64}, ::Diagonal{Float64, Vector{Float64}})`
probably it's fixable... orrr we could just drop 1.6 and only run 1 for tests. do you have a strong preference for continuing to test on 1.6?

**Comment by odow** [comment]
I have a strong preference to maintain testing on 1.6, but it's okay if you add a `if VERSION > v"1.6"` around that particular test?

**Comment by chriscoey** [comment]
Ok I'll just be less lazy and fix it

File: src/Cones/epitrrelentropytri.jl [hunk 1]
```diff
@@ -440,85 +455,109 @@ function Δ3!(Δ3::Array{T, 3}, Δ2::Matrix{T}, λ::Vector{T}) where {T <: Real}
...
+        mul!(mat3, Hermitian(mat2, :U), V)
+        mul!(mat2, V', mat3)
+        @views smat_to_svec!(d2zdV2[:, col_idx], mat2, rt2)
+        col_idx += 1
+    end
     return d2zdV2
 end
 
+# NOTE helper for hess_prod, totally untested, can have typos
+# function d2zdV2_prod!(
```
Comment (chriscoey) [inline src/Cones/epitrrelentropytri.jl:509]: @lkapelevich @araujoms what should we do with this commented-out hessian product related code? might be worth trying to get it working?

File: src/Cones/epitrrelentropytri.jl [hunk 2]
```diff
@@ -440,85 +455,109 @@ function Δ3!(Δ3::Array{T, 3}, Δ2::Matrix{T}, λ::Vector{T}) where {T <: Real}
...
+    d3WlogVdV::Matrix{R},
     Δ3::Array{T, 3},
     λ::Vector{T},
-    Vds::Matrix{T},
-    Ws::Matrix{T},
-    Δ4_ij::Matrix{T}, # temp
-) where {T <: Real}
+    Vds::Matrix{R},
+    Ws::Matrix{R},
+    Δ4_ij::Matrix{R}, # temp TODO could have been a Matrix{T}
```
Comment (chriscoey) [inline src/Cones/epitrrelentropytri.jl:554]: re the comment "TODO could have been a Matrix{T}", should we change that now?

File: src/Cones/epitrrelentropytri.jl (hunk 1 above)
Comment (chriscoey) [reply src/Cones/epitrrelentropytri.jl:509]: ah thanks and good point. OK i'll see if i can add a native test

Comment (chriscoey) [reply src/Cones/epitrrelentropytri.jl:509]: added native tests for the complex case

Comment (chriscoey) [reply src/Cones/epitrrelentropytri.jl]: thanks, just added your code in the latest commit.

**Comment by chriscoey** [comment]
I just pushed @lkapelevich's hessian product oracle code. @araujoms it might be worth just running your use cases again (assuming CI goes green) now before we decide to merge, to make sure the new code doesn't cause any regressions. Let us know if performance is good and then I think we can merge!

**Comment by chriscoey** [comment]
tests failed:
```
epitrrelentropytri1 Float64 : Error During Test at /home/runner/work/Hypatia.jl/Hypatia.jl/test/runnativetests.jl:38
  Got exception outside of a @test
  AssertionError: cone.hess_updated
  Stacktrace:
    [1] update_dder3_aux(cone::Hypatia.Cones.EpiTrRelEntropyTri{Float64, Float64})
      @ Hypatia.Cones ~/work/Hypatia.jl/Hypatia.jl/src/Cones/epitrrelentropytri.jl:363
    [2] dder3(cone::Hypatia.Cones.EpiTrRelEntropyTri{Float64, Float64}, dir::Vector{Float64})
      @ Hypatia.Cones ~/work/Hypatia.jl/Hypatia.jl/src/Cones/epitrrelentropytri.jl:373
    [3] update_rhs_centadj(solver::Hypatia.Solvers.Solver{Float64}, rhs::Hypatia.Solvers.Point{Float64}, dir::Hypatia.Solvers.Point{Float64})
      @ Hypatia.Solvers ~/work/Hypatia.jl/Hypatia.jl/src/Solvers/steppers/common.jl:103
```
so i'll have to try to fix that before you try running anything, @araujoms

**Comment by chriscoey** [comment]
@lkapelevich would you mind checking whether my last commit (a one line change to update hess aux instead of asserting the whole hessian is updated, for dder3) is what you intended? i ran the cone oracle and barrier tests and they passed.

**Comment by araujoms** [comment]
I've ran some test cases, and while the answers all came up correct, I believe there is a mistake somewhere. In the easy instances the version with the product oracle was much faster, but on the hard cases it caused numerical problems and became much slower. I can email you the code that triggers the issue if you want.

**Comment by chriscoey** [comment]
@araujoms thanks for testing. could you do so again? @lkapelevich identified a bug and we hope it's fixed now

**Comment by araujoms** [comment]
It definitely helped, the numerical instability and the massive slowdown are gone. The behaviour is still weird, though, in some hard instances using the product oracle is still much slower than not using it.

**Comment by chriscoey** [comment]
Hmm, @araujoms in the cases where it is worse now, does it take more iterations, or is each iteration just slower?

**Comment by araujoms** [comment]
The iterations are just slower, the total number is the same.

**Comment by chriscoey** [comment]
OK. the purpose of the hessian product oracle is mainly to improve numerical performance (when numerics get ugly). historically it has not caused a significant slowdown for other cones. it sounds like maybe your cases are numerically OK already. can you quantify us what sort of slowdown you are seeing? are the tests you're running in the Hypatia examples? if so then we could maybe try to see where the most time is spent in the hessian product oracle to see if there's some low hanging fruit.

**Comment by araujoms** [comment]
In the easy instances the product oracle works as expected, giving more than 5x speedup. In the hard cases, however, it causes a significant slowdown, taking about 50% longer in some cases. I suspect it will get worse for even harder cases, but I haven't tested it.

The code I'm running is unpublished research so I'm not going to make it public, but I can email it to you.

**Comment by chriscoey** [comment]
Hm I wouldn't expect such a speedup on small cases. Which two commits above are you comparing?

**Comment by araujoms** [comment]
The already-merged part of `complexquantum` versus the current PR.

**Comment by chriscoey** [comment]
No complex quantum cone PRs have been merged into Hypatia main branch though - everything needed to enable the complex case is on this PR. Here is the cone on main https://github.com/jump-dev/Hypatia.jl/blob/master/src/Cones/epitrrelentropytri.jl - only the real case is supported. There are a lot of changes in this PR beyond just the hessian product oracle.

**Comment by araujoms** [comment]
No no, I meant the part that is merged in `complexquantum` branch, not in master. There is a `complexquantum` branch in the Hypatia repository, which is the target of this PR.

**Comment by chriscoey** [comment]
I only see one complexquantum branch, this one. and the target of this PR is master: 
image

**Comment by araujoms** [comment]
Oh. Somewhere a mess was made, but there is still a `complexquantum` branch in the Hypatia repository, [here](https://github.com/jump-dev/Hypatia.jl/tree/complexquantum). I should have named the branch in my fork differently.

**Comment by chriscoey** [comment]
right but that branch is this PR 831's branch: 
image

**Comment by chriscoey** [comment]
Anyway, I think we should merge this as-is. Our default is to add a hessian product when we can do so. It has helped for numerics, and in some cases, speed. It has also caused some minor speed regressions in cases where numerics were already fine, but that hasn't stopped us in the past.

**Comment by araujoms** [comment]
No, this PR's branch lives in my fork. The latest commit in Hypatia's `complexquantum` branch is [aaf6669](https://github.com/jump-dev/Hypatia.jl/commit/aaf666991265b15620e172f6e7db5fa341835ba4), whereas the latest commit in this PR is [f08bfbf](https://github.com/jump-dev/Hypatia.jl/pull/831/commits/f08bfbf0a4d530df81092c8015c0806b922f090f).

In any case, I'm ok with merging, I can easily disable the product oracle in my code to do more careful benchmarks, and perhaps find what is causing the slowdown.

I just wanted your opinion on the first question I asked, is there a way to get rid of my hack and get `smat_to_svec!` to dispatch correctly on Hermitian JuMP variables?

Also, there is another puzzling behaviour that I have observed, at some point I inadvertently added a redundant PSD constraint for one of the Hermitian matrices going into the relative entropy cone. To my surprise this did not make the solver go crazy, but instead made it faster.

**Comment by araujoms** [comment]
Completely unrelated: I think equation (60) of `coneref.pdf` is incorrect, I checked the code and it's not interleaving the `v` and `w` vectors.

**Comment by chriscoey** [comment]
@araujoms thanks for the concrete analysis. could you try to narrow it down to the commit(s) that cause the performance degradation you see?

**Comment by araujoms** [comment]
I did a binary search and found the offending commit: [676b055](https://github.com/jump-dev/Hypatia.jl/pull/831/commits/676b05565855ab39e75261f81162f19b71f8c525). That's where I added the asymptotic expansion for the central point, which delivered a closer approximation to it.

I am very confused. I was certain that there would be no noticeable difference in performance, as the nonlinear fit you were using already delivered a good approximation, and the difference from mine was only about 10^-4. If anything I would expect the performance to be slightly better, not 2 times worse!

I've now tested providing an even better approximation, correct up to 10^-6. It did improve the performance with respect to my asymptotic expansion, but it was still significantly slower than your "bad" approximation.

I don't know what to do. If simply providing the best central point gave the optimal performance I could just generate the central points for dimensions 2 to 1000 and add that as a file, leaving the rest to the asymptotic expansion. But since this is not the case, how on Earth can I find the central point that gives the optimal performance!?

**Comment by araujoms** [comment]
The central points of the matrix relative entropy cone are a simple function of the central points of the vector relative entropy cone, so the code just calls them from there. I don't use the actual vector relative entropy cone for anything.

**Comment by chriscoey** [comment]
Got it, right. Sorry I'm on mobile so can't look more closely.

I think this situation is just a fluke. Something about your specific examples seems to make the old central point value work slightly better for larger cases. This kind of thing happens... Really we need a broad diversity of examples generated from many different problems to show that certain changes to Hypatia help or hinder performance on average. Lacking that, we just use simple math heuristics and some guesswork. 

I like that you derived asymptotic values for the central point (though I didn't verify them), and you can decide whether to leave them in the PR. Maybe you just want to increase the dimension threshold to something higher. In any case, modulo what you decide there, I think this PR is done.

**Comment by araujoms** [comment]
You're right, I've tested with a couple other problems, and the performance does indeed vary, sometimes it's better with a closer approximation to the central point, sometimes it's not.

Since the performance depends so critically on the central point I'd like to add the solutions for the first 1000 dimensions. I'm unsure about what to do about the file IO; normally I'd use JLD2, but I assume you're not keen on adding another dependency to Hypatia.

**Comment by chriscoey** [comment]
Our experience is that performance impacts of central point precision are fairly random, as long as it's somewhat close, hence why we used pretty simple approximations. I'm not keen for this cone to have 1000 points, and I don't think it's going to make any noticeable difference in general, so long as the approximations we have already are within the vague neighborhood of the central point. If you want, for your paper, you could experiment with that. But I'd like to keep the central points that we have in Hypatia simple and consistent across the different cones.

**Comment by araujoms** [comment]
How about 100 points then? I could just save the literals in `epirelentropy.jl`, no file IO needed.

**Comment by chriscoey** [comment]
Sorry to be a pain but we don't do that level of precision for other cones - we settled on the simple step-wise approximations that we use because getting more precise than that wasn't really helpful on our benchmarks in general, and allows us to keep the code concise and maintainable. Even if you could show convincing evidence that actually hard coding 100 points is better on a variety of benchmarks from several different applications, I'd still be a littttttle reluctant. Why not just go back to our original initial points that we used before that commit that made things slower?

**Comment by chriscoey** [comment]
Also the initial point should (I hope) be user-specifiable without modifying Hypatia source code - so you could drop down to that in the experiments you are running without needing to fork Hypatia.

**Comment by araujoms** [comment]
Fine, I just set the asymptotic expansion to kick in after dimension 300, where your nonlinear fit really breaks down. Its of little relevance in practice as dimension 300 is hardly solvable.

**Comment by chriscoey** [comment]
@lkapelevich @odow @araujoms would you like to review this PR before merge?

**Comment by odow** [comment]
I don't think I understand the code enough to competently review.

**Comment by chriscoey** [comment]
Merged. I'll tag today

**Comment by chriscoey** [comment]
Thanks everyone!

**Comment by chriscoey** [comment]
New version 0.8.0 is tagged

**Pull request merged**